consensus/tendermint/propose.go

```diff
@@ -0,0 +1,246 @@
+*/
+func (t *Tendermint[V, H, A]) line49WhenProposalIsReceived(p Proposal[V, H, A], precommitsForHR map[A][]Precommit[H,
+	A], vID H, validProposal bool, proposalFromProposer bool,
+) bool {
+	var precommits []Precommit[H, A]
+	var vals []A
+
+	for addr, valPrecommits := range precommitsForHR {
+		for _, p := range valPrecommits {
+			if *p.ID == vID {
+				precommits = append(precommits, p)
+				vals = append(vals, addr)
+			}
+		}
+	}
+
+	if validProposal && proposalFromProposer &&
+		t.validatorSetVotingPower(vals) >= q(t.validators.TotalVotingPower(p.H)) {
+		// After committing the block, how the new height and round is started needs to be coordinated
+		// with the synchronisation process.
+		t.blockchain.Commit(t.state.h, *p.Value, precommits)
+
+		t.messages.deleteHeightMessages(t.state.h)
+		t.state.h++
+		t.startRound(0)
+
+		return true
+	}
+	return false
+}
+
+/*
+Check the upon condition on line 22:
+
+	22: upon {PROPOSAL, h_p, round_p, v, nil} from proposer(h_p, round_p) while step_p = propose do
+	23: 	if valid(v) ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) then
+	24: 		broadcast {PREVOTE, h_p, round_p, id(v)}
+	25: 	else
+	26: 		broadcast {PREVOTE, h_p, round_p, nil}
+	27:		step_p ← prevote
+
+The implementation uses nil as -1 to avoid using int type.
+
+Since the value's id is expected to be unique the id can be used to compare the values.
+*/
+func (t *Tendermint[V, H, A]) line22(vr int, proposalFromProposer, validProposal bool, vID H) {
+	if vr == -1 && proposalFromProposer && t.state.s == propose {
+		vote := Prevote[H, A]{
+			H:      t.state.h,
+			R:      t.state.r,
+			ID:     nil,
+			Sender: t.nodeAddr,
+		}
+
+		if validProposal && (t.state.lockedRound == -1 || (*t.state.lockedValue).Hash() == vID) {
+			vote.ID = &vID
+		}
+
+		t.messages.addPrevote(vote)
+		t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+		t.state.s = prevote
+	}
+}
+
+/*
+Check the upon condition on line 28:
+
+	28: upon {PROPOSAL, h_p, round_p, v, vr} from proposer(h_p, round_p) AND 2f + 1 {PREVOTE,h_p, vr, id(v)} while
+		step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do
+	29: if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) then
+	30: 	broadcast {PREVOTE, hp, round_p, id(v)}
+	31: else
+	32:  	broadcast {PREVOTE, hp, round_p, nil}
+	33: step_p ← prevote
+
+Ideally, the condition on line 28 would be checked in a single if statement, however,
+this cannot be done because valid round needs to be non-nil before the prevotes are fetched.
+*/
+func (t *Tendermint[V, H, A]) line28WhenProposalIsReceived(p Proposal[V, H, A], vr int, proposalFromProposer bool,
+	vID H, validProposal bool,
+) {
+	if vr != -1 && proposalFromProposer && t.state.s == propose && vr >= 0 && vr < int(t.state.r) {
+		_, prevotesForHVr, _ := t.messages.allMessages(p.H, round(vr))
+
+		var vals []A
+		for addr, valPrevotes := range prevotesForHVr {
+			for _, p := range valPrevotes {
+				if *p.ID == vID {
+					vals = append(vals, addr)
+				}
+			}
+		}
```

Comment:
nit: I think we can have a function for these lines, so each line will have single level of abstraction.